File: C# OOP 22'/Polymorphism/Polymorphism - Exercise/02.VehiclesExtension/Models/Vehicle.cs
﻿namespace Vehicles.Models
{

    using Contracts;
    using Exceptions;


    public abstract class Vehicle : IVehicle
    {

        protected Vehicle(double fuelQuantity, double fuelConsumption,  double tankCapacity)
        {

            this.FuelQuantity = fuelQuantity;
            //the counsumption increses coused by using the AC
            this.FuelConsumption = fuelConsumption;
            this.TankCapacity = tankCapacity;

            if (this.FuelQuantity > this.TankCapacity)
                this.FuelQuantity = 0;
        }

        public double FuelQuantity { get; private set; }

        public double FuelConsumption { get; protected set; }

        public double TankCapacity { get; private set; }

        public virtual string Drive(double distance)
        {
            double neededFuel = distance * this.FuelConsumption;

            if (neededFuel > this.FuelQuantity)
            {
                //not enough fuel
                throw new InsufficientFuelException
                    (string.Format(ExceptionMessage.InsufficientFuelExceptionMessage, this.GetType().Name));
            }

            this.FuelQuantity -= neededFuel;

            return $"{this.GetType().Name} travelled {distance} km";
        }

        //public string DriveEmpty(double distance)
        //{
        //    double neededFuel = distance * this.FuelConsumption;

        //    if (neededFuel > this.FuelQuantity)
        //    {
        //        //not enough fuel
        //        throw new InsufficientFuelException
        //            (string.Format(ExceptionMessage.InsufficientFuelExceptionMessage, this.GetType().Name));
        //    }

        //    this.FuelQuantity -= neededFuel;

        //    return $"{this.GetType().Name} travelled {distance} km";
        //}

        public virtual void Refuel(double liters)
        {
            if (liters <= 0)
            {
                throw new WrongFuelInput(ExceptionMessage.WRONG_FUEL_INPUT);
            }

            if (this.FuelQuantity + liters > this.TankCapacity)
            {
                throw new OverfillingTheTank(
                    string.Format(ExceptionMessage.OVERFILLING_THE_TANK, liters));
            }

            this.FuelQuantity += liters;
        }

        public override string ToString()
        {
            return $"{this.GetType().Name}: {this.FuelQuantity:f2}";
        }
    }
}

File: Python/OOP/Polymorphism and Abstractions/Exercise/vehicle.py
from abc import ABC, abstractmethod


class Vehicle(ABC):
    def __init__(self, fuel_quantity, fuel_consumption):
        self.fuel_quantity = fuel_quantity
        self.fuel_consumption = fuel_consumption

    @abstractmethod
    def drive(self, distance):
        pass

    @abstractmethod
    def refuel(self, fuel: float):
        pass


class Car(Vehicle):
    def drive(self, distance):
        fuel_needed = distance * (self.fuel_consumption + 0.9)
        if fuel_needed <= self.fuel_quantity:
            self.fuel_quantity -= fuel_needed

    def refuel(self, fuel: float):
        self.fuel_quantity += fuel


class Truck(Vehicle):
    def refuel(self, fuel: float):
        self.fuel_quantity += fuel * 0.95

    def drive(self, distance):
        fuel_needed = distance * (self.fuel_consumption + 1.6)
        if fuel_needed <= self.fuel_quantity:
            self.fuel_quantity -= fuel_needed


car = Car(20, 5)
car.drive(3)
print(car.fuel_quantity)
car.refuel(10)
print(car.fuel_quantity)
print(f'===================================')
truck = Truck(100, 15)
truck.drive(5)
print(truck.fuel_quantity)
truck.refuel(50)
print(truck.fuel_quantity)
